bpf/parser_common.h

```diff
@@ -0,0 +1,265 @@
+/*
+ * Copyright (c) 2018 Nicira, Inc.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of version 2 of the GNU General Public
+ * License as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA
+ */
+#include "api.h"
+#include "helpers.h"
+
+#include <linux/if_ether.h>
+#include <linux/if_arp.h>
+#include <linux/if_vlan.h>
+#include <linux/ip.h>
+#include <linux/ipv6.h>
+#include <linux/tcp.h>
+#include <linux/udp.h>
+#include <linux/icmp.h>
+
+#define TCP_FLAGS_BE16(tp) (*(__be16 *)&tcp_flag_word(tp) & bpf_htons(0x0FFF))
+
+#ifndef ErrorCode
+#define ErrorCode
+enum ErrorCode {
+    ovs_no_error,
+    ovs_index_out_of_bounds,
+    ovs_out_of_packet,
+    ovs_header_too_long,
+    ovs_header_too_short,
+    ovs_unhandled_select,
+    ovs_checksum,
+    ovs_too_many_encap,
+    ovs_ipv6_disabled,
+};
+#endif
+
+/* Parse packet data: used both in TC and XDP
+ * Save result in map: percpu_headers
+ */
+static inline int
+PARSE_DATA(struct OVS_SK_BUFF *skb)
+{
+    struct ebpf_headers_t hdrs = {};
+    ovs_be16 eth_proto;
+    u32 ebpf_zero = 0;
+    int offset = 0;
+    u8 nw_proto = 0;
+    int err = 0;
+
+    /* Link Layer. */
+    if (OVS_LOAD_BYTES(skb, offset, &hdrs.ethernet, sizeof(hdrs.ethernet)) < 0) {
+        err = ovs_header_too_short;
+        printt("ERR: load byte %d\n", __LINE__);
+        goto end;
+    }
+    if (hdrs.ethernet.etherType == 0) {
+        printt("Layer 3 packet with eth_proto == 0, return TC_ACT_OK\n");
+        return TC_ACT_OK;
+    }
+
+    offset += sizeof(hdrs.ethernet);
+    hdrs.valid |= ETHER_VALID;
+
+    eth_proto = hdrs.ethernet.etherType;
+
+    if (eth_proto == bpf_htons(ETH_P_8021Q)){
+
+        /* The inner vlan, if exists, is VLAN 8021Q (0x8100)
+         * The outer vlan is at skb metadata, could be 8021Q or 8021AD
+         */
+        struct vlan_hdr { /* wired format */
+            ovs_be16 tci;
+            ovs_be16 ethertype;
+        } cvlan;
+
+        /* parse cvlan */
+        if (OVS_LOAD_BYTES(skb, offset - 2, &cvlan, sizeof(cvlan)) < 0) {
+            err = ovs_header_too_short;
+            printt("ERR: load byte %d\n", __LINE__);
+            goto end;
+        }
+        offset += sizeof(hdrs.cvlan);
+        hdrs.valid |= CVLAN_VALID;
```

Comment:
it should be inner vlan, so cvlan.
consider 2 cases:
* Packet has only 1 vlan case: 8021Q
XDP parse nothing, since it will be in skb metadata
* Packet has nested vlan (outer vlan 8021AD and inner 8021Q)
XDP parse inner 8021Q, and skb metadata has 8021AD

Reply:
My concern is that when XDP first receives a packet, the outer VLAN isn't parsed yet.

Reply:
yes, that's possible. So xdp will first parse inner vlan, and tc ebpf parses outer vlan from metadata.